### DISCUSSION TITLE
Using `next-intl` with app router without `middleware.ts`

### DISCUSSION THREAD

**Discussion by Ncookiez** [amannn/next-intl#975] [Q&A] [answered]
An unfortunate consequence of upgrading to the app router for unrelated reasons is that to use `next-intl` we seemingly need to use a `middleware.ts` file. This means NextJS 14 forces you to use an edge runtime, as opposed to the node runtime. I'll leave my rant about why this is horrendous for another time. Anyhow, my app is not compatible with this runtime, so...

**Can I still use `next-intl` without a `middleware.ts` file?**

I'm looking to only do basic locale detection/switching and load translations for each locale on pages/components, etc. If there is a limited approach I can take I'd love to hear about it, as I can't seem to find any in the docs or otherwise.

**Comment by amannn** [comment, marked as answer]
How are you hosting your app? E.g. if you self-host a Next.js app via `npm start`, the middleware will run locally in Node.js too. Only if you're deploying to a service like Vercel, then the middleware is executed in an edge runtime. Note, however, that your app can still run in Node.js, only the middleware is executed on the edge.

That being said, there is an option to [avoid using the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#usage-without-middleware-static-export) with `next-intl`.

Hope this helps!

**Reply by Ncookiez** [reply]
> if you self-host a Next.js app via npm start, the middleware will run locally in Node.js too

Unfortunately this doesn't seem to be the case. See [this RFC](https://github.com/vercel/next.js/discussions/46722) or [this issue](https://github.com/vercel/next.js/issues/48638).

Regardless, I appreciate the great docs link! I might've skipped it too hastily since I realistically can't statically export my app - do you think it'd be feasible to still use prefix-based routing without a static export or middleware? I imagine I'd need some redirects like the docs show, and probably detect the locale through a cookie/localstorage/header manually?

**Reply by amannn** [reply]
From [the Next.js deployment docs](https://nextjs.org/docs/app/building-your-application/deploying):
> When deploying using `next start`, all Next.js features are supported.

I'd give it a shot in your position, I've seen this work well for people who prefer self-hosting!

>  it'd be feasible to still use prefix-based routing without a static export or middleware?

Yep, absolutely—the docs mostly talk about this in regard to static export since this is a feature request that occasionally comes up, but you could still run your app with dynamic routes on a server.

**Reply by yordis** [reply]
@amannn I am not too concerned with self-hosting since I would use Vercel for deployment, but I would love to see the tool be as agnostic of Vercel infra as possible.

**Reply by amannn** [reply]
I'd argue it is, which concern do you have? Next.js is agnostic of Vercel, see [the deployment docs](https://nextjs.org/docs/app/building-your-application/deploying).

**Reply by m-sanders** [reply]
I am also self hosting and in a similar boat. The problem with using next middleware when self hosted is that the middleware sits behind the CDN, not at the edge. Round trips to the middleware for every route is expensive.
